File: monorepo_java/apps/web/src/main/java/mono/web/Features.java
package mono.web;

import java.util.List;
import java.util.ArrayList;

public class Features {
    public enum Priority {
        LOW(1), MEDIUM(5), HIGH(10);
        private final int weight;
        Priority(int weight) { this.weight = weight; }
        public int weight() { return weight; }
        public static Priority fromWeight(int w) {
            for (Priority p : values()) if (p.weight == w) return p;
            throw new IllegalArgumentException("bad weight " + w);
        }
    }

    public record Money(long amount, String currency) {
        public Money {
            if (currency == null) throw new IllegalArgumentException();
        }
        public Money add(Money other) {
            if (!currency.equals(other.currency)) throw new IllegalArgumentException();
            return new Money(amount + other.amount, currency);
        }
    }

    static double sumOfList(List<? extends Number> list) {
        double s = 0;
        for (Number n : list) s += n.doubleValue();
        return s;
    }

    static void fillWithZeros(List<? super Integer> list, int n) {
        for (int i = 0; i < n; i++) list.add(0);
    }

    public static class Resource implements AutoCloseable {
        private final String name;
        public Resource(String name) { this.name = name; System.out.println("open " + name); }
        public void use() { System.out.println("use " + name); }
        @Override public void close() { System.out.println("close " + name); }
    }

    static String describe(Priority p) {
        return switch (p) {
            case LOW -> "take your time";
            case MEDIUM -> "soon please";
            case HIGH -> "drop everything";
        };
    }

    public static void runDemo() {
        System.out.println("priority weight = " + Priority.HIGH.weight());
        System.out.println("from weight 5 = " + Priority.fromWeight(5));
        System.out.println("description = " + describe(Priority.MEDIUM));

        Money m1 = new Money(100, "USD");
        Money m2 = new Money(50, "USD");
        System.out.println("sum = " + m1.add(m2));

        List<Integer> ints = List.of(1, 2, 3);
        System.out.println("sum = " + sumOfList(ints));
        List<Number> nums = new ArrayList<>();
        fillWithZeros(nums, 3);
        System.out.println("nums = " + nums);

        try (Resource a = new Resource("a"); Resource b = new Resource("b")) {
            a.use();
            b.use();
        }
    }
}

File: monorepo_php/apps/web/src/Features.php
<?php
namespace Mono\Web;

// Ported from php/Features.php. PHP 8 attribute + enum + match + readonly.

#[\Attribute(\Attribute::TARGET_METHOD | \Attribute::TARGET_CLASS)]
class AuditedAttr {
    public function __construct(public readonly string $reason = "") {}
}

enum PriorityFeat: int {
    case Low    = 1;
    case Medium = 5;
    case High   = 10;

    public function describe(): string {
        return match($this) {
            PriorityFeat::Low    => "take your time",
            PriorityFeat::Medium => "soon please",
            PriorityFeat::High   => "drop everything",
        };
    }
}

final class MoneyFeat {
    public function __construct(
        public readonly int $amount,
        public readonly string $currency,
    ) {}

    public function add(MoneyFeat $other): MoneyFeat {
        if ($other->currency !== $this->currency) {
            throw new \InvalidArgumentException("currency mismatch");
        }
        return new MoneyFeat($this->amount + $other->amount, $this->currency);
    }
}

#[AuditedAttr(reason: "top-level service")]
class ServiceFeat {
    #[AuditedAttr(reason: "side-effecting")]
    public function run(PriorityFeat $p, MoneyFeat $m): string {
        $note = match(true) {
            $p === PriorityFeat::High && $m->amount > 1000 => "urgent big payment",
            $p === PriorityFeat::High                      => "urgent",
            default                                        => "normal",
        };
        return $p->describe() . " / " . $note;
    }
}

function phpFeaturesDemoMono(): void {
    $p = PriorityFeat::High;
    echo "priority weight = " . $p->value . PHP_EOL;
    echo "describe = " . $p->describe() . PHP_EOL;

    $m1 = new MoneyFeat(500, "USD");
    $m2 = new MoneyFeat(750, "USD");
    $sum = $m1->add($m2);
    echo "sum = {$sum->amount} {$sum->currency}" . PHP_EOL;

    $svc = new ServiceFeat();
    echo $svc->run(PriorityFeat::High, $sum) . PHP_EOL;

    $ref = new \ReflectionClass(ServiceFeat::class);
    foreach ($ref->getAttributes(AuditedAttr::class) as $attr) {
        echo "attr reason=" . $attr->newInstance()->reason . PHP_EOL;
    }
}
